File: Python/remove-outermost-parentheses.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def removeOuterParentheses(self, S):
        """
        :type S: str
        :rtype: str
        """
        deep = 1
        result, cnt = [], 0
        for c in S:
            if c == '(' and cnt >= deep:
                result.append(c)
            if c == ')' and cnt > deep:
                result.append(c)
            cnt += 1 if c == '(' else -1
        return "".join(result)

File: C++/remove-outermost-parentheses.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    string removeOuterParentheses(string S) {
        static const int deep = 1;
        string result;
        int cnt = 0;
        for (const auto& c : S) {
            if (c == '(' && cnt++ >= deep) {
                result += c;
            }
            if (c == ')' && cnt-- > deep) {
                result += c;
            }
        }
        return result;
    }
};
